sockeye/loss.py

```diff
@@ -168,3 +169,38 @@ def get_loss(self, logits: mx.sym.Symbol, labels: mx.sym.Symbol) -> List[mx.sym.
         cross_entropy = mx.sym.MakeLoss(cross_entropy, name=C.SMOOTHED_CROSS_ENTROPY)
         probs = mx.sym.BlockGrad(probs, name=C.SOFTMAX_NAME)
         return [cross_entropy, probs]
+
+
+@register
+@alias("sce")
+class PreComputedSmoothedCrossEntropy(EvalMetric):
+    """
+    This is a simple helper class that can be called as a metric to sum pre-computed smoothed
+    cross entropy values.  It was written mainly as a bridge to make SCE loss available to extended
+    optimizers during training.
+
+    :param normalized: Whether loss is pre-normalized by the number of examples.
+    :param name: Name of this metric instance for display.
+    :param output_names: Name of predictions that should be used when updating with update_dict.
+    :param output_labels: Name of labels that should be used when updating with update_dict.
+    """
+    def __init__(self,
+                 normalized: bool = False,
+                 name: str = "smoothed-cross-entropy",
+                 output_names: Optional[List[str]] = None,
+                 label_names: Optional[List[str]] = None) -> None:
+        super().__init__(name, output_names=output_names, label_names=label_names)
+        self.normalized = normalized
+
+    def update(self, labels, preds):
+        """Sum pre-computed smoothed cross entropy values"""
+        # Take cross entropy values only
+        sces = preds[::2]
+        for label, sce in zip(labels, sces):
+            label = label.asnumpy()
+            label = label.ravel()
+            assert label.shape[0] == sce.shape[0]
+            # SCE is pre-computed, so just sum
+            self.sum_metric += sce.asnumpy().sum()
```

Comment:
we could do the sum in mx.nd space and only copy to numpy afterwards. If the sces live on the gpu this is cheaper and faster.
Same above with the ravel. You can ask ndarrays for their shape as well and this is faster than copying to numpy before.

Reply:
Another case where code borrowed from MXNet's implementation isn't as efficient as it could be.